## CARGAR LOS DATOS

Importamos las librerias numpy (operaciones matemáticas), pandas (dataframes), seaborn y matplotlib(visualización).

In [1]:
import numpy as np 
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

Leemos el archivo .csv original y creamos una copia.

In [2]:
dataset = pd.read_csv("../data/GBvideos_cc50_202101.csv")
datasetcopia = dataset

## INSPECCIONAR LOS DATOS

Exploramos el dataframe, con la instrucción *info()*

In [3]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38916 entries, 0 to 38915
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   video_id                38916 non-null  object 
 1   trending_date           38916 non-null  object 
 2   title                   38916 non-null  object 
 3   channel_title           38916 non-null  object 
 4   category_id             38916 non-null  int64  
 5   publish_time            38916 non-null  object 
 6   tags                    38916 non-null  object 
 7   views                   38916 non-null  int64  
 8   likes                   38916 non-null  int64  
 9   dislikes                38916 non-null  int64  
 10  comment_count           38916 non-null  int64  
 11  thumbnail_link          38916 non-null  object 
 12  comments_disabled       38916 non-null  bool   
 13  ratings_disabled        38916 non-null  bool   
 14  video_error_or_removed  38916 non-null

Dado esta salida, sabemos que existen 38916 entradas (filas), con 19 variables (columnas)

Además. inspeccionamos con la función *describe()*

In [6]:
dataset.describe()

category_id         views         likes      dislikes  comment_count  \
count  38916.000000  3.891600e+04  3.891600e+04  3.891600e+04   3.891600e+04   
mean      16.827937  5.911944e+06  1.345196e+05  7.612560e+03   1.308835e+04   
std        7.752728  1.900121e+07  3.499893e+05  5.095683e+04   5.066740e+04   
min        1.000000  8.510000e+02  0.000000e+00  0.000000e+00   0.000000e+00   
25%       10.000000  2.515272e+05  5.897000e+03  2.000000e+02   6.790000e+02   
50%       20.000000  9.818890e+05  2.518250e+04  8.210000e+02   2.478000e+03   
75%       24.000000  3.683628e+06  1.140892e+05  3.357500e+03   9.241500e+03   
max       43.000000  4.245389e+08  5.613827e+06  1.944971e+06   1.626501e+06   

                lat           lon  
count  38916.000000  38916.000000  
mean      53.383720     -2.379078  
std        2.043531      2.103878  
min       50.133722     -7.333284  
25%       51.499995     -3.529950  
50%       53.200020     -2.180007  
75%       54.879995     -1.080022  
max       58.966981      1.300013

Ahora, veamos las entradas con la instrucción *head()*

In [7]:
dataset.head()

video_id trending_date  \
0  Jw1Y-zhQURU      17.14.11   
1  3s1rvMFUweQ      17.14.11   
2  n1WpP7iowLc      17.14.11   
3  PUTEiSjKwJU      17.14.11   
4  rHwDegptbI4      17.14.11   

                                               title  \
0      John Lewis Christmas Ad 2017 - #MozTheMonster   
1          Taylor Swift: …Ready for It? (Live) - SNL   
2         Eminem - Walk On Water (Audio) ft. Beyoncé   
3  Goals from Salford City vs Class of 92 and Fri...   
4  Dashcam captures truck's near miss with child ...   

                channel_title  category_id              publish_time  \
0                  John Lewis           26  2017-11-10T07:38:29.000Z   
1         Saturday Night Live           24  2017-11-12T06:24:44.000Z   
2                  EminemVEVO           10  2017-11-10T17:00:03.000Z   
3  Salford City Football Club           17  2017-11-13T02:30:38.000Z   
4            Cute Girl Videos           25  2017-11-13T01:45:13.000Z   

                                                tags     views   likes  \
0  christmas|"john lewis christmas"|"john lewis"|...   7224515   55681   
1  SNL|"Saturday Night Live"|"SNL Season 43"|"Epi...   1053632   25561   
2  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579  787420   
3  Salford City FC|"Salford City"|"Salford"|"Clas...     27833     193   
4                                             [none]      9815      30   

   dislikes  comment_count                                  thumbnail_link  \
0     10247           9479  https://i.ytimg.com/vi/Jw1Y-zhQURU/default.jpg   
1      2294           2757  https://i.ytimg.com/vi/3s1rvMFUweQ/default.jpg   
2     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
3        12             37  https://i.ytimg.com/vi/PUTEiSjKwJU/default.jpg   
4         2             30  https://i.ytimg.com/vi/rHwDegptbI4/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description  \
0  Click here to continue the story and make your...   
1  Musical guest Taylor Swift performs …Ready for...   
2  Eminem's new track Walk on Water ft. Beyoncé i...   
3  Salford drew 4-4 against the Class of 92 and F...   
4  Dashcam captures truck's near miss with child ...   

                          state        lat       lon  \
0                         Devon  50.700405 -3.529950   
1                    Manchester  53.500415 -2.247987   
2  Bath And North East Somerset  51.383749 -2.350022   
3                      Aberdeen  57.170398 -2.079987   
4                   Southampton  50.900031 -1.399977   

                           geometry  
0  POINT (-3.529950197 50.70040529)  
1  POINT (-2.247987103 53.50041526)  
2   POINT (-2.350022218 51.3837486)  
3  POINT (-2.079987021 57.17039797)  
4  POINT (-1.399976849 50.90003135)

## PRE-PROCESAR LOS DATOS

### VALORES FALTANTES

Vamos a inspeccionar cuantos NAs existen en cada columna

In [8]:
datasetcopia.isna().sum()

video_id                    0
trending_date               0
title                       0
channel_title               0
category_id                 0
publish_time                0
tags                        0
views                       0
likes                       0
dislikes                    0
comment_count               0
thumbnail_link              0
comments_disabled           0
ratings_disabled            0
video_error_or_removed      0
description               612
state                       0
lat                         0
lon                         0
geometry                    0
dtype: int64

Podemos notar que solo, el description tiene 612 valore NAs. Entonces, podemos asignar su valor con 'No description'.

In [9]:
datasetcopia['description'].fillna('No description', inplace=True)

Revisamos la cantidad de NA's para confirmar.

In [11]:
datasetcopia.isna().sum()

video_id                  0
trending_date             0
title                     0
channel_title             0
category_id               0
publish_time              0
tags                      0
views                     0
likes                     0
dislikes                  0
comment_count             0
thumbnail_link            0
comments_disabled         0
ratings_disabled          0
video_error_or_removed    0
description               0
state                     0
lat                       0
lon                       0
geometry                  0
dtype: int64

### OUTLIERS

Creamos una función para arreglar los outliers

In [12]:
def arreglarOutliers(data):
    outliers = data.quantile([0.05,0.95])
    n = len(data)
    outliers = list(outliers)
    mean = data.mean()
    median = data.median()
    
    for element in range(n):
        if int(data[element]) < int(outliers[0]):
            data.loc[element]= mean
        elif int(data[element]) > int(outliers[1]):
            data.loc[element]= median

    return data

Outliers de la columnas 'views'

### TRENDING_DATE, PUBLISH_TIME

In [13]:
datasetcopia['trending_date'] = pd.to_datetime(datasetcopia['trending_date'], format='%y.%d.%m')
datasetcopia['publish_time'] = pd.to_datetime(datasetcopia['publish_time'])

### ARREGLAR OUTLIERS DE LA COLUMNA FECHA

In [14]:
def arreglarOutliersfecha(data):
    outliers = data.quantile([0.02,0.98])
    n = len(data)
    outliers = list(outliers)
    for element in range(n):
        if data[element] > outliers[1] or data[element]<outliers[0]:
            data[element] = np.nan

### CREAR COLUMNA PARA DESCRIPCIÓN DE LA CATEGORIA

In [15]:
dfCategory = pd.read_json('GB_category_id.json')

In [16]:
dfCategory.head()

kind  \
0  youtube#videoCategoryListResponse   
1  youtube#videoCategoryListResponse   
2  youtube#videoCategoryListResponse   
3  youtube#videoCategoryListResponse   
4  youtube#videoCategoryListResponse   

                                                etag  \
0  "m2yskBQFythfE4irbTIeOgYYfBU/1v2mrzYSYG6onNLt2...   
1  "m2yskBQFythfE4irbTIeOgYYfBU/1v2mrzYSYG6onNLt2...   
2  "m2yskBQFythfE4irbTIeOgYYfBU/1v2mrzYSYG6onNLt2...   
3  "m2yskBQFythfE4irbTIeOgYYfBU/1v2mrzYSYG6onNLt2...   
4  "m2yskBQFythfE4irbTIeOgYYfBU/1v2mrzYSYG6onNLt2...   

                                               items  
0  {'kind': 'youtube#videoCategory', 'etag': '"m2...  
1  {'kind': 'youtube#videoCategory', 'etag': '"m2...  
2  {'kind': 'youtube#videoCategory', 'etag': '"m2...  
3  {'kind': 'youtube#videoCategory', 'etag': '"m2...  
4  {'kind': 'youtube#videoCategory', 'etag': '"m2...

In [17]:
dfCategory.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32 entries, 0 to 31
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   kind    32 non-null     object
 1   etag    32 non-null     object
 2   items   32 non-null     object
dtypes: object(3)
memory usage: 448.0+ bytes


In [18]:
items = dfCategory['items']

In [19]:
category = dict()
for i in range(len(dfCategory)):
    category[items[i]['id']] = items[i]['snippet']['title']

In [20]:
datasetcopia['category_description'] = [ category[str(i)] for i in datasetcopia['category_id']]
datasetcopia['category_description']

0          Howto & Style
1          Entertainment
2                  Music
3                 Sports
4        News & Politics
              ...       
38911              Music
38912              Music
38913              Music
38914      Entertainment
38915              Music
Name: category_description, Length: 38916, dtype: object

### EXPORT CSV

In [22]:
datasetcopia.to_csv('../data/GB_dataset_preprocessed.csv', index=False)

In [23]:
data = pd.read_csv('../data/GB_dataset_preprocessed.csv')
data

video_id trending_date  \
0      Jw1Y-zhQURU    2017-11-14   
1      3s1rvMFUweQ    2017-11-14   
2      n1WpP7iowLc    2017-11-14   
3      PUTEiSjKwJU    2017-11-14   
4      rHwDegptbI4    2017-11-14   
...            ...           ...   
38911  l884wKofd54    2018-06-14   
38912  IP8k2xkhOdI    2018-06-14   
38913  Il-an3K9pjg    2018-06-14   
38914  -DRsfNObKIQ    2018-06-14   
38915  4YFo4bdMO8Q    2018-06-14   

                                                   title  \
0          John Lewis Christmas Ad 2017 - #MozTheMonster   
1              Taylor Swift: …Ready for It? (Live) - SNL   
2             Eminem - Walk On Water (Audio) ft. Beyoncé   
3      Goals from Salford City vs Class of 92 and Fri...   
4      Dashcam captures truck's near miss with child ...   
...                                                  ...   
38911  Enrique Iglesias - MOVE TO MIAMI (Official Vid...   
38912  Jacob Sartorius - Up With It (Official Music V...   
38913                 Anne-Marie - 2002 [Official Video]   
38914  Eleni Foureira - Fuego - Cyprus - LIVE - First...   
38915  KYLE - Ikuyo feat.  2 Chainz & Sophia Black [A...   

                    channel_title  category_id               publish_time  \
0                      John Lewis           26  2017-11-10 07:38:29+00:00   
1             Saturday Night Live           24  2017-11-12 06:24:44+00:00   
2                      EminemVEVO           10  2017-11-10 17:00:03+00:00   
3      Salford City Football Club           17  2017-11-13 02:30:38+00:00   
4                Cute Girl Videos           25  2017-11-13 01:45:13+00:00   
...                           ...          ...                        ...   
38911         EnriqueIglesiasVEVO           10  2018-05-09 07:00:01+00:00   
38912             Jacob Sartorius           10  2018-05-11 17:09:16+00:00   
38913                  Anne-Marie           10  2018-05-08 11:05:08+00:00   
38914     Eurovision Song Contest           24  2018-05-08 20:32:32+00:00   
38915              SuperDuperKyle           10  2018-05-11 04:06:35+00:00   

                                                    tags     views   likes  \
0      christmas|"john lewis christmas"|"john lewis"|...   7224515   55681   
1      SNL|"Saturday Night Live"|"SNL Season 43"|"Epi...   1053632   25561   
2      Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579  787420   
3      Salford City FC|"Salford City"|"Salford"|"Clas...     27833     193   
4                                                 [none]      9815      30   
...                                                  ...       ...     ...   
38911  Enrique Iglesias feat. Pitbull|"MOVE TO MIAMI"...  25066952  268088   
38912  jacob sartorius|"jacob"|"up with it"|"jacob sa...   1492219   61998   
38913  anne|"marie"|"anne-marie"|"2002"|"two thousand...  29641412  394830   
38914  Eurovision Song Contest|"2018"|"Lisbon"|"Cypru...  14317515  151870   
38915  Kyle|"SuperDuperKyle"|"Ikuyo"|"2 Chainz"|"Soph...    607552   18271   

       dislikes  ...                                  thumbnail_link  \
0         10247  ...  https://i.ytimg.com/vi/Jw1Y-zhQURU/default.jpg   
1          2294  ...  https://i.ytimg.com/vi/3s1rvMFUweQ/default.jpg   
2         43420  ...  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
3            12  ...  https://i.ytimg.com/vi/PUTEiSjKwJU/default.jpg   
4             2  ...  https://i.ytimg.com/vi/rHwDegptbI4/default.jpg   
...         ...  ...                                             ...   
38911     12783  ...  https://i.ytimg.com/vi/l884wKofd54/default.jpg   
38912     13781  ...  https://i.ytimg.com/vi/IP8k2xkhOdI/default.jpg   
38913      8892  ...  https://i.ytimg.com/vi/Il-an3K9pjg/default.jpg   
38914     45875  ...  https://i.ytimg.com/vi/-DRsfNObKIQ/default.jpg   
38915       274  ...  https://i.ytimg.com/vi/4YFo4bdMO8Q/default.jpg   

      comments_disabled  ratings_disabled  video_error_or_removed  \
0                 False             False                   

## Nuevo data frame con las frecuencias

In [34]:
data1 = data.set_index(['trending_date', 'video_id'])
data1.head()

title  \
trending_date video_id                                                         
2017-11-14    Jw1Y-zhQURU      John Lewis Christmas Ad 2017 - #MozTheMonster   
              3s1rvMFUweQ          Taylor Swift: …Ready for It? (Live) - SNL   
              n1WpP7iowLc         Eminem - Walk On Water (Audio) ft. Beyoncé   
              PUTEiSjKwJU  Goals from Salford City vs Class of 92 and Fri...   
              rHwDegptbI4  Dashcam captures truck's near miss with child ...   

                                        channel_title  category_id  \
trending_date video_id                                               
2017-11-14    Jw1Y-zhQURU                  John Lewis           26   
              3s1rvMFUweQ         Saturday Night Live           24   
              n1WpP7iowLc                  EminemVEVO           10   
              PUTEiSjKwJU  Salford City Football Club           17   
              rHwDegptbI4            Cute Girl Videos           25   

                                        publish_time  \
trending_date video_id                                 
2017-11-14    Jw1Y-zhQURU  2017-11-10 07:38:29+00:00   
              3s1rvMFUweQ  2017-11-12 06:24:44+00:00   
              n1WpP7iowLc  2017-11-10 17:00:03+00:00   
              PUTEiSjKwJU  2017-11-13 02:30:38+00:00   
              rHwDegptbI4  2017-11-13 01:45:13+00:00   

                                                                        tags  \
trending_date video_id                                                         
2017-11-14    Jw1Y-zhQURU  christmas|"john lewis christmas"|"john lewis"|...   
              3s1rvMFUweQ  SNL|"Saturday Night Live"|"SNL Season 43"|"Epi...   
              n1WpP7iowLc  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...   
              PUTEiSjKwJU  Salford City FC|"Salford City"|"Salford"|"Clas...   
              rHwDegptbI4                                             [none]   

                              views   likes  dislikes  comment_count  \
trending_date video_id                                                 
2017-11-14    Jw1Y-zhQURU   7224515   55681     10247           9479   
              3s1rvMFUweQ   1053632   25561      2294           2757   
              n1WpP7iowLc  17158579  787420     43420         125882   
              PUTEiSjKwJU     27833     193        12             37   
              rHwDegptbI4      9815      30         2             30   

                                                           thumbnail_link  \
trending_date video_id                                                      
2017-11-14    Jw1Y-zhQURU  https://i.ytimg.com/vi/Jw1Y-zhQURU/default.jpg   
              3s1rvMFUweQ  https://i.ytimg.com/vi/3s1rvMFUweQ/default.jpg   
              n1WpP7iowLc  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
              PUTEiSjKwJU  https://i.ytimg.com/vi/PUTEiSjKwJU/default.jpg   
              rHwDegptbI4  https://i.ytimg.com/vi/rHwDegptbI4/default.jpg   

                           comments_disabled  ratings_disabled  \
trending_date video_id                                           
2017-11-14    Jw1Y-zhQURU              False             False   
              3s1rvMFUweQ              False             False   
              n1WpP7iowLc              False             False   
              PUTEiSjKwJU              False             False   
              rHwDegptbI4              False             False   

                           video_error_or_removed  \
trending_date video_id                              
2017-11-14    Jw1Y-zhQURU                   False   
              3s1rvMFUweQ                   False   
              n1WpP7iowLc                   False   
              PUTEiSjKwJU                   False   
              rHwDegptbI4                   False   

                                                                 description  \
trending_date video_id                                                         
2017-11-14    Jw1Y-zhQURU

In [41]:
# Nuevo dataframe con Video Level Stats
video_level = data1.groupby(level=1).last()
video_level['freq'] = data1['title'].groupby(level=1).count()
video_level.sort_values(by=['views'], ascending=False, inplace=True)

video_level.head(10)

title  \
video_id                                                         
_I_D_8Z4sJE  Nicky Jam x J. Balvin - X (EQUIS) | Video Ofic...   
9jI-z9QN6g8  Te Bote Remix - Casper, Nio García, Darell, Ni...   
kLpH1nSLJSs               Bad Bunny - Amorfoda | Video Oficial   
wfWkmURBNv8           Ozuna x Romeo Santos - El Farsante Remix   
VYOjWnS4cMY  Childish Gambino - This Is America (Official V...   
xpVfcZ0ZcFM                                 Drake - God’s Plan   
ffxKSjUwKdU               Ariana Grande - No Tears Left To Cry   
zEf423kYfqk  Becky G, Natti Natasha - Sin Pijama (Official ...   
FlsCjmMhFmw  YouTube Rewind: The Shape of 2017 | #YouTubeRe...   
sGIm0-dQd8M                Dura - Daddy Yankee (Video Oficial)   

                   channel_title  category_id               publish_time  \
video_id                                                                   
_I_D_8Z4sJE           NickyJamTV           10  2018-03-02 05:00:19+00:00   
9jI-z9QN6g8        Flow La Movie           10  2018-04-11 22:00:00+00:00   
kLpH1nSLJSs            Bad Bunny           10  2018-02-15 00:00:03+00:00   
wfWkmURBNv8                Ozuna           10  2018-01-30 15:00:05+00:00   
VYOjWnS4cMY  ChildishGambinoVEVO           10  2018-05-06 04:00:07+00:00   
xpVfcZ0ZcFM            DrakeVEVO           10  2018-02-17 05:00:01+00:00   
ffxKSjUwKdU     ArianaGrandeVevo           10  2018-04-20 04:00:03+00:00   
zEf423kYfqk           BeckyGVEVO           10  2018-04-20 10:40:51+00:00   
FlsCjmMhFmw    YouTube Spotlight           24  2017-12-06 17:58:51+00:00   
sGIm0-dQd8M         Daddy Yankee           10  2018-01-18 22:32:49+00:00   

                                                          tags      views  \
video_id                                                                    
_I_D_8Z4sJE  Bad Bunny|"Amorfoda"|"Dura"|"Daddy Yankee"|"Oz...  424538912   
9jI-z9QN6g8  Te Bote|"Te Bote Remix"|"Flow La Movie"|"Ozuna...  337621571   
kLpH1nSLJSs  Bad|"Bunny"|"Bad Bunny"|"Video Oficial"|"Music...  328860380   
wfWkmURBNv8  Ozuna|"Reggaeton"|"Odisea"|"Odisea The Album"|...  288811992   
VYOjWnS4cMY  Childish Gambino|"Rap"|"This Is America"|"mcDJ...  259721696   
xpVfcZ0ZcFM  Drake new music|"Drake Gods Plan"|"Drake God’s...  258164991   
ffxKSjUwKdU  Ariana|"Grande"|"No"|"Tears"|"Left"|"To"|"Cry"...  208876887   
zEf423kYfqk  Becky G|"Natti Natasha"|"Natti Natasha Music"|...  200862743   
FlsCjmMhFmw  Rewind|"Rewind 2017"|"youtube rewind 2017"|"#Y...  169884583   
sGIm0-dQd8M  daddy yankee reggaeton|"daddy yankee youtube"|...  167456025   

               likes  dislikes  comment_count  \
video_id                                        
_I_D_8Z4sJE  2818771    149275          99380   
9jI-z9QN6g8  2581961    166549         113564   
kLpH1nSLJSs  3823879    215530         225216   
wfWkmURBNv8  1618180    151147          67506   
VYOjWnS4cMY  5444541    379862         553371   
xpVfcZ0ZcFM  4737873    117198         301756   
ffxKSjUwKdU  3394437    150086         259613   
zEf423kYfqk  1668418    142569          97826   
FlsCjmMhFmw  3312868   1753274         845233   
sGIm0-dQd8M  1633407     74005          65395   

                                             thumbnail_link  \
video_id                                                      
_I_D_8Z4sJE  https://i.ytimg.com/vi/_I_D_8Z4sJE/default.jpg   
9jI-z9QN6g8  https://i.ytimg.com/vi/9jI-z9QN6g8/default.jpg   
kLpH1nSLJSs  https://i.ytimg.com/vi/kLpH1nSLJSs/default.jpg   
wfWkmURBNv8  https://i.ytimg.com/vi/wfWkmURBNv8/default.jpg   
VYOjWnS4cMY  https://i.ytimg.com/vi/VYOjWnS4cMY/default.jpg   
xpVfcZ0ZcFM  https://i.ytimg.com/vi/xpVfcZ0ZcFM/default.jpg   
ffxKSjUwKdU  https://i.ytimg.com/vi/ffxKSjUwKdU/default.jpg   
zEf423kYfqk  https://i.ytimg.com/vi/zEf423kYfqk/default.jpg   
FlsCjmMhFmw  https://i.ytimg.com/vi/FlsCjmMhFmw/default.jpg   
sGIm0-dQd8M  https://i.ytimg.com/vi/sGIm0-dQd8M/default.jpg   

             comments_disabled  ratings_disabled  video_error_or_removed  \
vide

## Preguntas

### 1. ¿Qué categorías de videos son las de mayor tendencia